### PR TITLE
bmat: add faster boolean matrices

### PR DESCRIPTION
**Pull request by james-d-mitchell**
This PR adds the class `BMat8` which provides faster boolean matrices.

### CODE REVIEW

File: src/bmat.cc [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+// This file contains an implementation of fast boolean matrices up to
+// dimension 8 x 8.
+
+#include "bmat.h"
+
+#include <algorithm>
```
Comment (flsmith) [inline src/bmat.cc]: I don't think this include is needed

File: tests/bmat.test.cc [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+             {1, 1, 1, 1, 1, 1, 1, 1},
+             {1, 1, 1, 1, 1, 1, 1, 1},
+             {1, 1, 1, 1, 1, 1, 1, 1},
+             {1, 1, 0, 1, 1, 1, 1, 1},
+             {0, 1, 1, 1, 0, 1, 1, 1},
+             {0, 1, 1, 1, 0, 1, 1, 1},
+             {1, 1, 1, 1, 1, 1, 1, 1}});
+
+  REQUIRE(tmp == bm3);
+  REQUIRE(tmp == bm * bm2);
```
Comment (flsmith) [inline tests/bmat.test.cc]: Should probably have more tests for multiplication, but I can't see a decent / non-trivial way of testing a lot of multiplications directly. The row space basis test did include a large number of multiplications.

File: Makefile.am [hunk 3]
```diff
@@ -76,15 +76,15 @@ CHECK_FLAGS2  = [quick],[standard]
 
 check_PROGRAMS = lstest
 
-lstest_SOURCES =  tests/main.test.cc
+lstest_SOURCES =  tests/main.test.cc      tests/bmat.test.cc
```
Comment (flsmith) [inline Makefile.am]: This isn't in the correct place. Do you care?

Comment (james-d-mitchell) [reply Makefile.am]: Yes, probably

Comment (james-d-mitchell) [reply Makefile.am]: Oh wait you mean `main.test.cc` is in the wrong place, I think we should leave it there

File: src/bmat.cc (hunk 1 above)
Comment (james-d-mitchell) [reply src/bmat.cc]: Will remove

File: tests/bmat.test.cc (hunk 2 above)
Comment (james-d-mitchell) [reply tests/bmat.test.cc]: Probably but I am in the process of adding tests for in the `template` branch that will perform millions of multiplications, so maybe don't worry about this for now.

File: Makefile.am (hunk 3 above)
Comment (flsmith) [reply Makefile.am]: I agree with leaving the ``main.test.cc`` where it is, but then the ``bmat.test.cc`` is still placed inconsistently relative to the others

**Comment by james-d-mitchell** [comment]
Superceded by #71

**Pull request closed without merging**